### PR TITLE
Fix issues on CI about guest-pull

### PR DESCRIPTION
**Pull request by ChengyuZhu6**
Let me explain why:

In our previous approach, we implemented guest pull by passing PullImageRequest to the guest. However, this method  resulted in the loss of specifications essential for running the container, such as commands specified in YAML, during the CreateContainer stage. To address this, it is necessary to integrate the OCI specifications and process information from the image’s configuration with the container in guest pull.

The snapshotter method does not care this issue. Nevertheless, a problem arises when two containers in the same pod attempt to pull the same image, like InitContainer. This is because the image service searches for the existing configuration, which resides in the guest. The configuration, associated with <image name, cid>, is stored in the directory `/run/kata-containers/<cid>`. Consequently, when the InitContainer finishes its task and terminates, the directory ceases to exist. As a result, during the creation of the application container, the OCI spec and process information cannot be merged due to the absence of the expected configuration file.

The root cause of the failure in k8s-kill-all-process-in-container.bats is the missing patch that exists only in the CCv0 branch: https://github.com/kata-containers/kata-containers/pull/2636/commits/44ed3ab80ee13413815d07f6a80a6e92fe3e8a93

Fix tests:
- k8s-credentials-secrets.bats
- k8s-file-volume.bats
- k8s-nested-configmap-secret.bats
- k8s-projected-volume.bats
- k8s-volume.bats
- k8s-shared-volume.bats 
- k8s-sysctls.bats
- k8s-inotify.bats
- k8s-liveness-probes.bats
- k8s-kill-all-process-in-container.bats
Fixes: #9664 #9665 #9666 #9667 #9668

### CODE REVIEW

**Comment by fidencio** [comment]
@ChengyuZhu6, I've re-enabled all the tests, so let's ensure those will pass.

File: src/agent/src/image.rs
Comment (stevenhorsman) [inline src/agent/src/image.rs:1]: Hey @ChengyuZhu6 - could you add some of the explanation from the issue into the commit message please?

**Comment by wainersm** [comment]
Hi @ChengyuZhu6 !

I gave the fix a try on my environment but I am still getting `k8s-empty-dirs.bats`, `k8s-kill-all-process-in-container.bats`  and `k8s-shared-volume.bats` failing. The `k8s-kill-all-process-in-container.bats` test in particular is supposed to be fixed as it launches an initContainer.

Comparing with the tests that failed in PR https://github.com/kata-containers/kata-containers/pull/9694 , this change made `k8s-file-volume.bats`, `k8s-seccomp.bats` and `k8s-sysctls.bats` to pass. Note that `k8s-seccomp.bats` is a case where you must run multiple times until you get the error, i.e., the fact that it passed on the tests triggered on this PR don't tell much. I'm afraid it's the case of other tests.

BTW, I'm using the image ghcr.io/kata-containers/kata-deploy-ci:9695-adf7662e0b8120fda812a8a375ef23ebae552b00-amd64 that was built on this PR. Also using `qemu-coco-dev` configuration with the changes as in https://github.com/kata-containers/kata-containers/pull/9694

**Comment by ryansavino** [comment]
With the commit to replace the docker container with the quay busybox, we are only seeing these three tests failing for snp:
```
k8s-kill-all-process-in-container.bats
k8s-sysctls.bats
k8s-shared-volume.bats
```

See #9678. @AdithyaKrishnan please update with the SEV results.

**Comment by ryansavino** [comment]
I think there is a problem with nydus. I'm seeing a bunch of other tests failing now for snp with messages like so for several different images:

```
Events:
  Type     Reason     Age               From               Message
  ----     ------     ----              ----               -------
  Normal   Scheduled  72s               default-scheduler  Successfully assigned kata-containers-k8s-tests/sharevol-kata to kata-ci-amd-milan-001
  Normal   Pulled     6s (x6 over 60s)  kubelet            Container image "registry.k8s.io/e2e-test-images/agnhost:2.21" already present on machine
  Warning  Failed     6s (x6 over 60s)  kubelet            Error: failed to create containerd container: error unpacking image: failed to extract layer sha256:721384ec99e56bc06202a738722bcb4b8254b9bbd71c43ab7ad0d9e773ced7ac: failed to get reader from content store: content digest sha256:5a3ea8efae5d0abb93d2a04be0a4870087042b8ecab8001f613cdc2a9440616a: not found
  Normal   Pulled     6s (x6 over 60s)  kubelet            Container image "registry.k8s.io/e2e-test-images/agnhost:2.21" already present on machine
  Warning  Failed     6s (x6 over 60s)  kubelet            Error: failed to create containerd container: error unpacking image: failed to extract layer sha256:721384ec99e56bc06202a738722bcb4b8254b9bbd71c43ab7ad0d9e773ced7ac: failed to get reader from content store: content digest sha256:5a3ea8efae5d0abb93d2a04be0a4870087042b8ecab8001f613cdc2a9440616a: not found
```

I think when nydus gets cleaned up, there is still some remnants of the old images laying about somewhere; the manifest has a record of it, but it's not on disk because it's getting cleaned up at the end of the test run.

I will continue to look into things and update with any more info.

**Comment by ChengyuZhu6** [comment]
https://github.com/kata-containers/kata-containers/actions/runs/9388686372/job/25856116319?pr=9695
I think most of tests are fixed except k8s-kill-all-process-in-container.bats.

**Review by arronwy — APPROVED** [review]
Thanks @ChengyuZhu6 LGTM!

**Comment by arronwy** [comment]
Hi @fidencio @stevenhorsman @fitzthum we verified this PR locally and online CI, now with guest pull feature enabled, we can still pass all kata Kubernetes integration tests! Please help review.

**Comment by ChengyuZhu6** [comment]
> Hi @fidencio @stevenhorsman @fitzthum we verified this PR locally and online CI, now with guest pull feature enabled, we can still pass all kata Kubernetes integration tests! Please help review.

Tests have passed on SEV and SNP
https://github.com/kata-containers/kata-containers/actions/runs/9789614780/job/27030498179?pr=9695
https://github.com/kata-containers/kata-containers/actions/runs/9789614780/job/27030500257?pr=9695

**Comment by ChengyuZhu6** [comment]
Rebase to main, no code change.

**Review by stevenhorsman — COMMENTED** [review]
I think the code looks good, but that we might need to re-work some of the commit messages to help with clarity, for example the message in d08a0488a0035ad8736ae2f65b4ea0a7b7a5a0b9 contain info about all the tests fixes in the PR and the github issues that are addressed in the whole PR. It might be clearer to to this to split that out of this commit and just address the double mount issue that the code in the commit is doing, then in the later commits that re-enable the tests we can group them into the tests that are mentioned in the kata-containers issues and add the fixes there?

Maybe someone less close to this like @gkurz could comment if this would be helpful, or if it's clear enough at the moment?

**Comment by ChengyuZhu6** [comment]
> I think the code looks good, but that we might need to re-work some of the commit messages to help with clarity, for example the message in [d08a048](https://github.com/kata-containers/kata-containers/commit/d08a0488a0035ad8736ae2f65b4ea0a7b7a5a0b9) contain info about all the tests fixes in the PR and the github issues that are addressed in the whole PR. It might be clearer to to this to split that out of this commit and just address the double mount issue that the code in the commit is doing, then in the later commits that re-enable the tests we can group them into the tests that are mentioned in the kata-containers issues and add the fixes there?
> 
> Maybe someone less close to this like @gkurz could comment if this would be helpful, or if it's clear enough at the moment?

Yeah, thanks for your advice. I have updated the commits.

File: src/agent/src/image.rs [hunk 1]
```diff
@@ -68,7 +68,7 @@ impl ImageService {
     }
 
     /// pause image is packaged in rootfs
-    fn unpack_pause_image(cid: &str, target_subpath: &str) -> Result<String> {
+    fn unpack_pause_image(cid: &str) -> Result<String> {
```
[the remark this answers is not in the document]
Comment (ChengyuZhu6) [reply src/agent/src/image.rs:63]: Thanks, I have split the commit.

File: src/agent/src/image.rs [hunk 2]
```diff
@@ -102,9 +102,7 @@ impl ImageService {
             bail!("The number of args should be greater than or equal to one! Please check the pause image.");
         }
 
-        let container_bundle = scoped_join(CONTAINER_BASE, cid)?;
-        fs::create_dir_all(&container_bundle)?;
-        let pause_bundle = scoped_join(&container_bundle, target_subpath)?;
+        let pause_bundle = scoped_join(CONTAINER_BASE, cid)?;
```
[the remark this answers is not in the document]
Comment (ChengyuZhu6) [reply src/agent/src/image.rs:97]: Done.

File: tests/integration/kubernetes/k8s-file-volume.bats [hunk 3]
```diff
@@ -12,8 +12,6 @@ TEST_INITRD="${TEST_INITRD:-no}"
 setup() {
 	[ "${KATA_HYPERVISOR}" == "firecracker" ] && skip "test not working see: ${fc_limitations}"
 	[ "${KATA_HYPERVISOR}" == "fc" ] && skip "test not working see: ${fc_limitations}"
-	[[ "${KATA_HYPERVISOR}" == "qemu-tdx" || "${KATA_HYPERVISOR}" == "qemu-coco-dev" ]] && \
-		skip "See: https://github.com/kata-containers/kata-containers/issues/9667"
```
[the remark this answers is not in the document]
Comment (ChengyuZhu6) [reply tests/integration/kubernetes/k8s-file-volume.bats:15]: Thanks. I have made the changes as your suggestion.

**Comment by stevenhorsman** [comment]
I think we have an issue with a job rename here, so I'll need to close and re-open this to refresh the required list :(

**Comment by ChengyuZhu6** [comment]
Remove the issue https://github.com/kata-containers/kata-containers/issues/2617 in the cherry-picked commit to fix ci https://github.com/kata-containers/kata-containers/actions/runs/9797575567/job/27063344127?pr=9695

**Comment by ChengyuZhu6** [comment]
Tests have passed on TDX.

**Pull request merged**